crates/matrix-sdk-ui/src/timeline/inner/mod.rs

```diff
@@ -607,24 +607,34 @@ impl<P: RoomDataProvider> TimelineInner<P> {
         }
     }
 
-    /// Handle a back-paginated event.
+    /// Handle a list of back-paginated events.
     ///
-    /// Returns the number of timeline updates that were made.
+    /// Returns the number of timeline updates that were made. Short-circuits
+    /// and returns `None` if the number of items added or updated exceeds
+    /// `u16::MAX`, which should practically never happen.
     #[instrument(skip_all)]
-    pub(super) async fn handle_back_paginated_event(
+    pub(super) async fn handle_back_paginated_events(
         &self,
-        event: TimelineEvent,
-    ) -> HandleEventResult {
-        self.state
-            .lock()
-            .await
-            .handle_remote_event(
-                event.into(),
-                TimelineItemPosition::Start,
-                &self.room_data_provider,
-                &self.settings,
-            )
-            .await
+        events: Vec<TimelineEvent>,
```

Comment:
nit: could this take an `&[TimelineEvent]`, or do we need ownership in there? (The one call site that pass a `vec![event]` would avoid an allocation with that change)

Reply:
It would be less efficient to not take ownership of the events. The other call site is test code, that we run without optimizations so perf will by abysmal anyways.